### PR TITLE
WIP: feat: `rubble-nrf51`/nRF51 support

### LINKED ISSUE #25: Support nRF51/thumbv6 targets (again)
Thumbv6 support was (accidentally) dropped when pulling in bbqueue (also see https://github.com/jamesmunns/bbqueue/issues/25). We need to fix that issue or stop using bbqueue, and then add a thumbv6 build job to CI.

This is needed for nRF51 support, since those all have a Cortex-M0.

### PR DESCRIPTION
**Pull request by chocol4te**
Closes #25.

Following must be complete before we can merge:

- [ ] https://github.com/jamesmunns/bbqueue/pull/27 needs to be merged, and the git dependency removed
- [x] New `log` release so we can remove that git dependency
- [ ] Build this configuration in CI
- [ ] Test connecting and enumerating services on real hardware
- [ ] Add nrf52810 support to `rubble-demo`

### CODE REVIEW

**Comment by chocol4te** [comment]
Not yet ready, need to fix the following:

 - [ ] https://github.com/jamesmunns/bbqueue/pull/27 needs to be merged, and the git dependancy removed
 - [ ] New `log` release so we can remove that git dependancy

File: rubble/src/lib.rs [hunk 1]
```diff
@@ -24,7 +24,7 @@
 #[macro_use]
 mod log;
 #[macro_use]
-mod utils;
+pub mod utils;
```
Comment (jonas-schievink) [inline rubble/src/lib.rs]: Why was this required?

File: rubble/Cargo.toml [hunk 2]
```diff
@@ -14,10 +14,13 @@ edition = "2018"
...
 
 # If the `log` feature is enabled, the `log` crate's macros will be called at various points to dump
 # packets, state, and events. By default, it is disabled.
 [dependencies.log]
-version = "0.4.6"
+git = "https://github.com/rust-lang-nursery/log"
 optional = true
+
+[features]
+cortex-m = ["bbqueue/cortex-m"]
```
Comment (jonas-schievink) [inline rubble/Cargo.toml]: Please add documentation to this features (in particular, mention that it must be enabled if Rubble is used on thumbv6 targets). Perhaps it's better to rename the feature to `thumbv6` for clarity?

File: rubble-nrf51/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+//! A Rubble BLE driver for the nRF52-series radios.
```
[the remark this answers is not in the document]
Comment (chocol4te) [reply rubble-nrf51/src/lib.rs]: Fixed :)

File: rubble-nrf51/src/timer.rs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+macro_rules! impl_timer {
+    ($ty:ident) => {
+        impl NrfTimerExt for $ty {
+            unsafe fn duplicate(&self) -> Self {
+                mem::transmute_copy(self)
+            }
+
+            fn init(&mut self) {
+                self.bitmode.write(|w| w.bitmode()._32bit());
```
[the remark this answers is not in the document]
Comment (chocol4te) [reply rubble-nrf51/src/timer.rs]: Fixed.

File: rubble/src/lib.rs (hunk 1 above)
Comment (chocol4te) [reply rubble/src/lib.rs]: I think I enabled it because I wanted to use hex slice in my application, I shouldn't have added that change. It does raise the question as to hexslice being split out as its own crate? It's tiny but pretty useful 🤷‍♀

Comment (jonas-schievink) [reply rubble/src/lib.rs]: I think `HexSlice` and `Hex` are the 2 types that I reimplement most often, especially in embedded projects, so yeah, maybe it would make sense to put them in their own crate. But just having these 2 in there is barely more than 20 lines of code, so that crate should probably do a bit more to be useful.

File: rubble-nrf51/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonas-schievink) [reply rubble-nrf51/src/lib.rs]: Still shows up as nRF52 though?

[the remark this answers is not in the document]
Comment (chocol4te) [reply rubble-nrf51/src/lib.rs]: Haven't pushed yet, fixed locally would be more accurate ;)

File: rubble/src/lib.rs (hunk 1 above)
Comment (chocol4te) [reply rubble/src/lib.rs]: That's what I was thinking, we don't want to end up with a million NPM-esque micro packages.

File: rubble/Cargo.toml (hunk 2 above)
Comment (chocol4te) [reply rubble/Cargo.toml]: Fixed (locally)

File: rubble/src/lib.rs (hunk 1 above)
Comment (chocol4te) [reply rubble/src/lib.rs]: Would it hurt to make the utils module public?

Comment (jonas-schievink) [reply rubble/src/lib.rs]: Maybe they can be added to the `hex` crate: https://docs.rs/hex

File: rubble-nrf51/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonas-schievink) [reply rubble-nrf51/src/lib.rs]: Oh okay. Seeing these comments is a bit confusing :)

File: rubble/src/lib.rs (hunk 1 above)
Comment (jonas-schievink) [reply rubble/src/lib.rs]: > Would it hurt to make the utils module public?

I don't really want to provide API stability for that module. It's meant to be an implementation detail of Rubble. If you need `HexSlice`, feel free to copy it out though :)

Comment (chocol4te) [reply rubble/src/lib.rs]: Ah good idea

File: rubble-nrf51/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (chocol4te) [reply rubble-nrf51/src/lib.rs]: Sorry, my bad

File: rubble/src/lib.rs (hunk 1 above)
Comment (jonas-schievink) [reply rubble/src/lib.rs]: They're already discussing something like this in https://github.com/KokaKiwi/rust-hex/issues/8

File: rubble-nrf51/src/timer.rs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+use {
+    core::mem,
+    nrf51_hal::nrf51::{TIMER0},
+    rubble::{
+        link::NextUpdate,
+        time::{Instant, Timer},
+    },
+};
+
+/// Implements Rubble's `Timer` trait for the timers on the nRF chip.
```
Comment (jonas-schievink) [inline rubble-nrf51/src/timer.rs]:
```suggestion
/// Implements Rubble's `Timer` trait for `TIMER0` on the nRF51 chips.
```

File: rubble-nrf51/src/timer.rs [hunk 6]
```diff
@@ -0,0 +1,176 @@
+//! Generic `Timer` implementation that works with TIMER0 on the chip.
```
Comment (jonas-schievink) [inline rubble-nrf51/src/timer.rs]:
```suggestion
//! Generic `Timer` implementation that works with `TIMER0` on the chip.
```

**Comment by jonas-schievink** [comment]
Another thing to do before merging: Test the demo on actual nRF51 hardware, making sure that connection and service enumeration works.

**Comment by chocol4te** [comment]
I'm sorting

> Another thing to do before merging: Test the demo on actual nRF51 hardware, making sure that connection and service enumeration works.

Doing that now, CI should also be testing that it builds on thumbv6 targets

**Comment by chocol4te** [comment]
So I modified the demo slightly for the nRF51, [rubble-demo-nrf51](https://github.com/chocol4te/rubble-demo-nrf51).

There appear to be 2 issues. 

The first is that logging does not appear to be working. Serial works, as `<<< INIT >>>` appears, but [this line is never called](https://github.com/chocol4te/rubble-demo-nrf51/blob/63781e31a672df2d21258d8f209d695fca30a201/src/main.rs#L216).

The second is that connecting hangs, but due to issue number 1, I don't really know how to go about diagnosing it.

IMG_0286

IMG_0287

**Comment by jonas-schievink** [comment]
`<< INIT >>` is printed via semihosting, while the log messages use the UART

**Comment by chocol4te** [comment]
Sorry I meant `--- INIT ---`, which gets printed after setting up the serial interface.

**Comment by jonas-schievink** [comment]
Maybe there's a bug in jamesmunns/bbqueue#27 that causes all data put in a bbqueue to get lost? Just a guess though.

**Comment by chocol4te** [comment]
Removed all the cfg's for logging, and now I get:

```
--- INIT ---
INFO - Logger ready
6.025ms -
```

Progress?

**Comment by jonas-schievink** [comment]
I'm pretty sure all those `Relaxed` orderings aren't strong enough. [`heapless`](https://github.com/japaric/heapless/blob/8ada965e274f0f00bf62bd9fcf034a50e847bbe8/src/sealed.rs#L58-L60) puts a `compiler_fence` with the correct ordering between all ops.

**Comment by chocol4te** [comment]
But `log::info!("Logger ready");` is working using the bbqueue? Or is it that one message could work, several would fail?

**Comment by chocol4te** [comment]
I'll make those changes in my fork and see what happens :)

**Comment by jonas-schievink** [comment]
Well, it would probably cause a data race and thus undefined behavior, so the program can do anything. How is the RAM usage? IIRC the log buffer is quite large. Do you have 32 or 16 KiB RAM on the chip? You could be running out of stack space.

**Comment by chocol4te** [comment]
The nRF51822 I have has 32K RAM, 256K flash. Maybe I should try that stack size tool?

**Comment by jonas-schievink** [comment]
No, the nRF52810 we were using only has 24K of RAM, so it can't be that

**Comment by chocol4te** [comment]
https://www.adafruit.com/product/2267

> As of July 29th, 2015 we're selling an updated version with a black PCB and the nRF51822 module with 32KB of SRAM.

**Comment by chocol4te** [comment]
Release build has interesting output:

```
Reading symbols from target/thumbv6m-none-eabi/release/rubble-demo-nrf51...
Target voltage: unknown
Available Targets:
No. Att Driver
 1      Nordic nRF51
0x0002d730 in ?? ()
Loading section .vector_table, size 0xa8 lma 0x0
Loading section .text, size 0xd97a lma 0xa8
Loading section .rodata, size 0x55c8 lma 0xda40
Loading section .data, size 0xc8 lma 0x13008
Start address 0xd9c8, load size 78002
Transfer rate: 26 KB/sec, 951 bytes/write.

<< INIT >>

panicked at 'called `Result::unwrap()` on an `Err` value: WouldBlock', src/libcore/result.rs:997:5

Program received signal SIGTRAP, Trace/breakpoint trap.
0x0000ac6c in __bkpt ()
(gdb)
```

**Comment by chocol4te** [comment]
```
<< INIT >>

panicked at 'there is no such thing as an acquire/release load', src/libcore/sync/atomic.rs:2127:19
```

????

https://doc.rust-lang.org/core/sync/atomic/enum.Ordering.html#variant.AcqRel

> Has the effects of both Acquire and Release together: For loads it uses Acquire ordering. For stores it uses the Release ordering.

**Comment by jonas-schievink** [comment]
> This ordering is only applicable for operations that combine both loads and stores.

**Comment by chocol4te** [comment]
🤦‍♂

**Comment by chocol4te** [comment]
```
--- INIT ---
INFO - Logger ready
6.037ms -
```
 Putting compiler fences everywhere doesn't appeared to have worked :(

**Comment by jonas-schievink** [comment]
Hmm, then I'm not sure what could cause this. I do remember seeing this myself, however (on a nRF52810). Not sure why it happened or how it got fixed.

**Comment by jonas-schievink** [comment]
The advertisement is showing up continuously though? Then there's definitely something wrong with the log queue not being filled/drained properly (once it's full it should panic). Or is the UART going silent?

**Comment by chocol4te** [comment]
Updated to [log](https://github.com/rust-lang-nursery/log) 0.4.7 after Jonas's changes were merged :)

**Comment by jonas-schievink** [comment]
Superseded by https://github.com/jonas-schievink/rubble/pull/97
